### ISSUE TITLE
I don't wanna be proxied by Washington

### ISSUE THREAD

**Issue by Karmavil** [microsoft/vscode-remote-release#5080]
🚨 The issue tracker is not for questions 🚨
If you have a question about VS Code, please ask it on https://stackoverflow.com/questions/tagged/visual-studio-code and about
VS Code Remote on https://stackoverflow.com/questions/tagged/vscode-remote.
___

This is a mix of question and complain about why vscode RC is exposing my dev environment to my network instead of exposing it just to the host. And yes the title is a bit sensationalist.

#### Why is Remote-Container "masking" or hiding to `docker ps` the assigned ports? 

and most important:

## Why is Microsoft proxying us?

Judging by the title, this issue is not strictly related: [#123037](https://github.com/microsoft/vscode/issues/123037#)
But believe me it is..
Someone (or something) was requesting a GET resource (probably knowing the _possible paths_ my dev backend returns to path-not-found) and it wasn't me (nor my app) and that's the last reply (today), when I started that issue (days ago) I was trying to report a lot of unusual amount of network traffic, so I am not sure, if this is malware, or intruders.
Internet connection was terrible this month ant the ISP was having troubles.. but I think someone cracked the wifi password to be honest. But then checking the network traffic I found that all packages downloaded when the container is build are download from the same IP
What's the reason for that?
And why VScodeRC is exposing the dev environment to my network? I cannot trust in my neighbors 🗡️ 

![proxied](https://user-images.githubusercontent.com/2354060/118940407-101aff80-b927-11eb-83a9-0ad9808cc9b4.png)

**Comment by Karmavil** [comment]
(I am not using an image from microsoft in my Dockerfile)

**Comment by Karmavil** [comment]
OMG was it tagged by the image? Nice! Great work

**Comment by Karmavil** [comment]
Thank you for not rejecting the issue, I understand it is now formatted as it should.
I appreciate it

**Comment by PavelSosin-320** [comment]
Calm, please! It can be plain CDN like Akamay that buffers static content. For the statically published content it generally improves performance. Software packages are, normally, downloaded from mirrors. You can check on which mirrors are published. The best way to avoid password cracking is frequent rotating like IP rotating. Adjust your "lease" time.

**Comment by chrmarti** [comment]
@Karmavil Addressing several issues:
- Your screenshot: Could you check what the full hostname of that IP address is? Do you see these with all extension disabled? With only the Remote-Containers extension enabled? Also check you have all 'online services' settings disabled. You can filter for those settings in the settings UI:
![image](https://user-images.githubusercontent.com/9205389/119126959-ffa47b00-ba33-11eb-8250-9eeb117f93df.png)
- On `docker ps`: We use `docker exec` to create a connection to the VS Code server running inside the container. Any ports being forwarded are tunneled through that `docker exec` connection. Are you expecting these ports to be shown with `docker ps`?
- We appear to be listening on all interfaces for the forwarded ports. Tracking this as #124350.

**Comment by Karmavil** [comment]
Many thanks for your replies:
@PavelSosin-320 oh no I am calm, there is someone in my network but that is all.. I'm nobody and there is nothing to stole from me. You are right, I have to update pwd more regularly. The password is complex but it is also complex how many users join/use the wifi (big family).
I can tell when it is a CDN because I see the prefix, like: cdn-185-199-110-1:https or I can search for it.

@chrmarti 
With **All plugins disabled** the connection to 40.64.128.224:https always happen regardless my Telemetry setting is unchecked.
Unchecked after your suggestion for this test specifically:
 *  Extensions: from *All Extensions* =to=> *None*
   * Auto Update 
   * Auto Check Updates
 * Update: 
   * Mode: from *default* =to=> *none* (restarted) 
   * Show Release Notes
 * Settings Sync: Keybindings Per Platform
 * JSON: Schema Download
 * Npm: Fetch Online Package Info

I'm wondering why I am seeing this csp1.zte.com.cn:domain
do you know what csp1 means? (that's out of the scope sorry) my host randomly stablish a connection with them
Ok I set the browser to off-line mode


All this is the result:
* You are right, no traffic at all to Microsoft servers after updates were disabled
* I could verify the traffic is to cloudflarenet when downloading images/packages(?) It should show me docker hub but .. anyway I'm not being proxied. At least when I build my own images, that it is what you reply to being responsible for Remote-Container not for VScode.

Al right, THANKS! Problem solved. Wise tips suggestion and procedure thank you 
So every time a vscode window is open a update check happens (or something like that)

**Comment by Karmavil** [comment]
Although the solution to the problem is partial, since I still cannot limit the scope of my development environment to local only, I understand that the solution would require a much more concise definition of the problem.

And I had having troubles trying to explain the problem, but I have to move on.

The problem when I declare ports manually in the docker-compose.yml is that they get unreachable from the browser, the only way I can connect to them is asking to vscode to publish a port for me to connect to the app, and this are different ports than the ones specified in the yaml file.

But that's a question suitable for stackoverflow I guess given that a viable workaround is avoid using the remote-container solutions and use my own yaml file instead. 
Maybe it is possible to declare those ports in a different way like "hacking" the docker excec chrmarti mentioned before I don't know.
That's all for now, thanks

**Comment by Karmavil** [comment]
OMG all this re-structure for nothing.. it was obvious...
The main differences between running my containers and running a container launched by remote container are not the networks, nor the port assignment.. it is what @chrmarti mention before the docker exec ... the way I am using remote container.
All I had to do was rename the yml to yaml for easy docker-compose up and down, declare the ports I need and networks.
Launch the container, attach vscode with remote container and that is it..
Finally! a simple solution and it was always there, now remote container is not responsible for the container

**Comment by Karmavil** [comment]
Now I feel terrible as a developer.
The real one thing that's needed is forward a port to the static port. 🙄
Anyway.. I learned something after all

**Comment by Karmavil** [comment]
> Now I feel terrible as a developer.
> The real one thing that's needed is forward a port to the static port. 🙄
> Anyway.. I learned something after all

Forget that.. it keeps exposing the listening port beyond my host
I'm starting to think that I'm wrong about what I think I know..  that's not good

**Comment by Karmavil** [comment]
Noup..  I am not wrong

Having a container with a server listening at port X in Dockerfile mapped to localhost:X in docker-compose file the limits are clear: you get a Err_connection_refuse trying to access from an external device.

* If you attach VScode or a shell and try to use the port the error is clear: **Error: That port is already in use.**
* But if the port is available and you use it to publish the app then the browser returns `The connection was reset`
Or, using the terminal: `curl: (56) Recv failure: Connection reset by peer`.

The way I made this works is with my own networks and ports declaration mapped to localhost, not start listening in the Dockerfile while in development, and in the container publish against 0.0.0.0
That keeps the isolation

I remember I figured this out months ago and I completely forgot it.

Conclusion: (IMHO) Port forwarding is as safe as your wifi password if your server dumps sensible data like credentials you're in troubles.
